psst-core/src/session/login5.rs

```diff
@@ -0,0 +1,269 @@
+            }
+            Login5Error::NoStoredCredentials => {
+                write!(f, "Tried to acquire access token without stored credentials")
+            }
+            Login5Error::RetriesFailed(u8) => {
+                write!(f, "Couldn't successfully authenticate after {:?} times", u8)
+            }
+        }
+    }
+}
+
+impl From<Login5Error> for Error {
+    fn from(err: Login5Error) -> Self {
+        match err {
+            Login5Error::NoStoredCredentials | Login5Error::RetriesFailed(_) | Login5Error::FaultyRequest(_) => {
+                Error::InvalidStateError(err.into())
+            }
+            Login5Error::CodeChallenge => Error::UnimplementedError(err.into()),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct Login5Request {
+    pub uri: String,
+    pub method: String,
+    pub payload: Vec<Vec<u8>>,
+}
+
+pub struct Login5 {
+    auth_token: Mutex<Option<Token>>,
+    client_token_provider: ClientTokenProvider,
+    agent: ureq::Agent,
+}
+
+impl Login5 {
+    /// Login5 instances can be used to cache and retrieve access tokens from stored credentials.
+    ///
+    /// # Arguments
+    ///
+    /// * `client_token_provider`: Can be optionally injected to control which client-id is
+    ///   used for it.
+    ///
+    /// returns: Login5
+    pub fn new(client_token_provider: Option<ClientTokenProvider>, proxy_url: Option<&str>) -> Self {
+        Self {
+            auth_token: Mutex::new(None),
+            client_token_provider: client_token_provider.unwrap_or_else(|| ClientTokenProvider::new(proxy_url)),
+            agent: default_ureq_agent_builder(proxy_url).build().into(),
+        }
+    }
+
+    fn request(&self, message: &LoginRequest) -> Result<Vec<u8>, Error> {
+        let client_token: String = self.client_token_provider.get()?;
+        let body = message.write_to_bytes()?;
+
+        let mut response = self
+            .agent
+            .post("https://login5.spotify.com/v3/login")
+            .header("Accept", "application/x-protobuf")
+            .header("client-token", &client_token)
+            .send(body)?;
+
+        let vec = response.body_mut().read_to_vec()?;
+        Ok(vec)
+    }
+
+    fn request_new_access_token(&self, login: Login_method) -> Result<Token, Error> {
+        let mut login_request = LoginRequest {
+            client_info: MessageField::some(ClientInfo {
+                client_id: String::from(CLIENT_ID),
+                device_id: String::from(DEVICE_ID),
+                special_fields: Default::default(),
+            }),
+            login_method: Some(login),
+            ..Default::default()
+        };
+
+        let mut response = self.request(&login_request)?;
+        let mut count = 0;
+
+        loop {
+            count += 1;
+
+            let message = LoginResponse::parse_from_bytes(&response)?;
+            if let Some(Response::Ok(ok)) = message.response {
+                return Ok(Token {
```

Comment:
For this and others below, we could use a match and match on Ok, error, challenges, etc. This would be a little cleaner.